xtask/src/task_phd.rs

```diff
@@ -95,6 +85,163 @@ pub(crate) struct RunArgs {
     /// `phd-runner`.
     #[clap(trailing_var_arg = true, allow_hyphen_values = true)]
     phd_args: Vec<String>,
+
+    #[clap(flatten)]
+    propolis_args: PropolisArgs,
+
+    #[clap(flatten)]
+    artifact_args: ArtifactStoreArgs,
+
+    #[clap(flatten)]
+    base_propolis_args: BasePropolisArgs,
+
+    #[clap(flatten)]
+    crucible_args: CrucibleArgs,
+}
+
+#[derive(Debug, Clone, clap::Parser)]
+#[group(id = "propolis", required = false, multiple = false)]
+#[command(next_help_heading = "Propolis Selection")]
+struct PropolisArgs {
+    /// The command to use to launch the Propolis server.
+    ///
+    /// If this is not present, a Propolis server binary will be built automatically.
+    #[clap(long = "propolis-server-cmd", value_parser, value_hint = clap::ValueHint::FilePath)]
+    server_cmd: Option<Utf8PathBuf>,
+
+    /// If set, build `propolis-server` in release mode.
+    #[clap(long, short = 'r')]
+    release: bool,
+}
+
+#[derive(Debug, Clone, clap::Parser)]
+#[group(id = "base-propolis", required = false, multiple = false)]
+#[command(next_help_heading = "Migration Base Propolis Selection")]
+struct BasePropolisArgs {
+    /// Git branch name to use for the "migration base" Propolis server artifact
+    /// for migration-from-base tests.
+    ///
+    /// If this argument is provided, PHD will download the latest Propolis
+    /// server artifact from Buildomat for the provided branch name, and use it
+    /// to test migration from that Propolis version to the Propolis revision
+    /// under test.
+    ///
+    /// This argument conflicts with the `--base-propolis-commit`,
+    /// `--base-propolis-cmd`, and `--no-base-propolis` arguments. If none of
+    /// these arguments are provided, `cargo xtask phd` will automatically pass
+    /// `--base-propolis-branch master` to `phd-runner`.
```

Comment:
If I'm reading this right the behavior of `BasePropolisArgs::configure_command` is to take the first matching option from (cmd, commit, branch, no_base) and ignore the others. Would it make sense to tell `clap` these options all conflict? (This applies to the `CrucibleArgs` as well I think.)

Reply:
This struct, as well as the `CrucibleArgs` and `PropolisArgs`, have a `clap` #[group]` attribute with [`multiple = false`](https://docs.rs/clap/latest/clap/struct.ArgGroup.html#method.multiple) on them: https://github.com/oxidecomputer/propolis/blob/66f09845d10a4c5c5b6ba1ae1cc4f23e1390376f/xtask/src/task_phd.rs#L118

This is the same as adding `conflicts_with` for each arg in the group to each individual arg. So, this validation already works:
```console
$ cargo xtask phd run --base-propolis-branch master --base-propolis-commit asdf --no-base-propolis

error: the argument '--base-propolis-branch <BASE_PROPOLIS_BRANCH>' cannot be used with:
  --base-propolis-commit <BASE_PROPOLIS_COMMIT>
  --no-base-propolis

Usage: xtask phd run --base-propolis-branch <BASE_PROPOLIS_BRANCH> [PHD_ARGS]...

For more information, try '--help'.
```

Reply:
Nice!